src/qibo/models/qcnn.py

```diff
@@ -0,0 +1,299 @@
+        expanded_params = []
+        nbits = self.nqubits
+        for layer in range(self.nlayers):
+            nleft = nbits / (2**layer)
+            param_start = layer * self.nparams_layer
+            conv_params = params[param_start : param_start + self.nparams_conv]
+            pool_params = params[
+                param_start + self.nparams_conv : param_start + self.nparams_layer
+            ]
+            # pool_params += pool_params[2::-1]
+            pool_params += [-pool_params[2], -pool_params[1], -pool_params[0]]
+            expanded_params += conv_params * int(nleft if nleft > 2 else 1)
+            expanded_params += pool_params * int(nleft / 2)
+
+        self._circuit.set_parameters(expanded_params)
+
+    def Classifier_circuit(self, theta):
+        """
+        Args:
+            theta: list or numpy.array with the biases and the angles to be used in the circuit
+            nlayers: int number of layers of the varitional circuit ansatz
+            RY: if True, parameterized Rx,Rz,Rx gates are used in the circuit
+                if False, parameterized Ry gates are used in the circuit (default=False)
+        Returns:
+            Circuit implementing the variational ansatz for angles "theta"
+        """
+        bias = np.array(theta[0 : self.measured_qubits])
+        angles = theta[self.measured_qubits :]
+
+        self.set_circuit_params(angles)
+        return self._circuit
+
+    def Predictions(self, circuit, theta, init_state, nshots=10000):
+        """
+        Args:
+            theta: list or numpy.array with the biases to be used in the circuit
+            init_state: numpy.array with the quantum state to be classified
+            nshots: int number of runs of the circuit during the sampling process (default=10000)
+        Returns:
+            numpy.array() with predictions for each qubit, for the initial state
+        """
+        bias = np.array(theta[0 : self.measured_qubits])
+        circuit = circuit(init_state, nshots)
+        result = circuit.frequencies(binary=False)
+        prediction = np.zeros(self.measured_qubits)
+
+        for qubit in range(self.measured_qubits):
+            for clase in range(self.nclasses):
+                binary = bin(clase)[2:].zfill(self.measured_qubits)
+                prediction[qubit] += result[clase] * (1 - 2 * int(binary[-qubit - 1]))
+
+        return prediction / nshots + bias
+
+    def square_loss(self, labels, predictions):
+        """
+        Args:
+            labels: list or numpy.array with the qubit labels of the quantum states to be classified
+            predictions: list or numpy.array with the qubit predictions for the quantum states to be classified
+        Returns:
+            numpy.float32 with the value of the square-loss function
+        """
+        loss = 0
+        for l, p in zip(labels, predictions):
+            for qubit in range(self.measured_qubits):
+                loss += (l[qubit] - p[qubit]) ** 2
```

Comment:
Here you use an MSE: ere the labels associated with the training data supposed to be continuous? I am used to encoding classes in discrete variables or in the components of a $N_{classes}$ dimensional vector, so I am interested in understanding this choice.

Reply:
The labels are $N_{\textrm{measured qubits}}$ dimensional vectors with each component being -1 or 1. Currently MSE is chosen because the predictions are "continuous". This choice is largely inspired by the variational classifier example, but we could possibly provide user the choice to specify loss function in the future.

Reply:
Okay, I understand. Maybe it might be useful to use the sign of the expected value as a class predictor.
In any case, I do not think that this choice will have a considerable impact on the results. 

By the way, I was wondering if you had saved any results or applications of this model: so that I could compare the accuracy of this QCNN with other models (I am curious about this).